spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -0,0 +1,516 @@
+        // and each tranche token up for rollover (held by perp),
+        // Execute rollover and move on to the next pair.
+        uint256 totalPerpRolledOver = 0;
+        uint8 i = 0;
+        uint256 j = 0;
+        while (i < td.trancheCount && j < rolloverTokens.length) {
+            ITranche trancheIn = td.tranches[i];
+            IERC20Upgradeable tokenOut = rolloverTokens[j];
+
+            // Compute available tranche in
+            uint256 trancheInAmtAvailable = trancheIn.balanceOf(address(this));
+
+            // trancheIn tokens are exhausted
+            if (trancheInAmtAvailable == 0) {
+                // Rollover is a no-op, so skipping to next trancheIn
+                i++;
+                continue;
+            }
+
+            // compute available token out
+            uint256 tokenOutAmtAvailable = address(tokenOut) != address(0) ? tokenOut.balanceOf(perp_.reserve()) : 0;
+
+            // trancheIn tokens are NOT exhausted but tokenOut is exhausted
+            if (tokenOutAmtAvailable == 0) {
+                // Rollover is a no-op, so skipping to next tokenOut
+                j++;
+                continue;
+            }
+
+            // Preview rollover
+            IPerpetualTranche.RolloverPreview memory rd = perp_.computeRolloverAmt(
+                trancheIn,
+                tokenOut,
+                trancheInAmtAvailable,
+                tokenOutAmtAvailable
+            );
+
+            // trancheIn isn't accepted by perp, likely because yield=0
+            if (rd.perpRolloverAmt == 0) {
+                // Though `trancheInAmtAvailable` it is unusable so marking as exhausted
+                // Rollover is a no-op, so skipping to next trancheIn
+                i++;
+                continue;
+            }
+
+            // Perform rollover
+            _checkAndApproveMax(trancheIn, address(perp_), trancheInAmtAvailable);
+            perp_.rollover(trancheIn, tokenOut, trancheInAmtAvailable);
+
+            // sync holdings
+            _syncDeployedAsset(trancheIn);
+            if (tokenOut != underlying) {
+                _syncDeployedAsset(tokenOut);
+            }
+            _syncAsset(perp_);
+            _syncAsset(underlying);
+
+            // keep track of total amount rolled over
+            totalPerpRolledOver += rd.perpRolloverAmt;
+        }
+
+        return totalPerpRolledOver;
+    }
+
+    /// @notice Redeems the deployed tranche tokens for the underlying asset.
+    function _redeemTranches() private {
+        for (uint256 i = 0; i < _deployed.length(); i++) {
+            ITranche tranche = deployedAt(i);
+            IBondController bond = IBondController(tranche.bond());
+
+            // if bond has mature, redeem the tranche token
+            if (bond.timeToMaturity() == 0) {
+                if (!bond.isMature()) {
+                    bond.mature();
+                }
+                bond.redeemMature(address(tranche), tranche.balanceOf(address(this)));
```

Comment:
Something to be cognizant of is that this can revert when the entire supply is redeemed and brick our `recoverAndRedeploy` workflow.

```
https://github.com/buttonwood-protocol/tranche/blob/main/contracts/BondController.sol#L225
```
Options:
1) We could wrap around try-catch
2) Add logic to compute amounts discounting the min condition
3) Handle manually if we ever run into this (deposit small amount of funds into the bond)

Reply:
Oh, that's right, good point. Could happen on `deposit` too if we don't deploy enough. I would say (2) compute the amounts, but since `MINIMUM_VALID_DEBT` isn't exposed, (1) try-catch works